src/libraries/System.IO.FileSystem.AccessControl/tests/TempAclDirectory.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Security.AccessControl;
+using System.Security.Principal;
+
+namespace System.IO
+{
+    /// <summary>
+    /// Represents a temporary directory.
+    /// Disposing will recurse all files and directories inside it, ensure the
+    /// appropriate access control is set, then delete all of them.
+    /// </summary>
+    public sealed class TempAclDirectory : TempDirectory
+    {
+        protected override void DeleteDirectory()
+        {
+            try
+            {
+                var identity = new SecurityIdentifier(WellKnownSidType.BuiltinUsersSid, null);
+                var accessRule = new FileSystemAccessRule(identity, FileSystemRights.FullControl, AccessControlType.Allow);
+
+                foreach (string dirPath in Directory.EnumerateDirectories(Path, "*", SearchOption.AllDirectories))
+                {
+                    var dirInfo = new DirectoryInfo(dirPath);
+                    var dirSecurity = new DirectorySecurity(dirPath, AccessControlSections.Access);
+                    dirSecurity.AddAccessRule(accessRule);
+                    dirInfo.SetAccessControl(dirSecurity);
+                }
+
+                foreach (string filePath in Directory.EnumerateFiles(Path, "*", SearchOption.AllDirectories))
+                {
+                    var fileInfo = new FileInfo(filePath);
+                    var fileSecurity = new FileSecurity(filePath, AccessControlSections.Access);
+                    fileSecurity.AddAccessRule(accessRule);
+                    fileInfo.SetAccessControl(fileSecurity);
+                }
+
+                var rootDirInfo = new DirectoryInfo(Path);
+                var rootSecurity = new DirectorySecurity(Path, AccessControlSections.Access);
+                rootSecurity.AddAccessRule(accessRule);
+                rootDirInfo.SetAccessControl(rootSecurity);
+                rootDirInfo.Delete(recursive: true);
+            }
+            catch { /* Do not throw because we call this on finalize */ }
```

Comment:
> I made sure to avoid enumerating files and folders in DeleteDirectory because that action expects access permission. Instead, any files or folders that a unit test creates, are passed to lists in the TempAclDirectory class. If non empty, they get traversed so the files and dirs they contain get their permissions restored (no need to enumerate if I already know what to modify).

Have you tried my suggestion with `InheritanceFlags.ContainerInherit | InheritanceFlags.ObjectInherit`? I wonder if it would allow us to simplify it further (don't store the files, just apply full permission recursively for the root folder and then delete the root folder recursively).

Reply:
Can you please look again? I simplified the dispose logic.